### PR TITLE
fix: add support for starknet address

### PR DESCRIPTION
**Pull request by wa0x6e**
Fixes https://github.com/snapshot-labs/pitches/issues/83

This PR introduces some changes in order to accept starknet address as input (in `where` and `id` fields)

### Changes

- feat: add `starknetAddress` to `where` fields accepted values
- feat: allow passing an array to `where` values, so we can format using `evmAddress`, or `['evmAddress', 'starknetAddress']`
- refactor: rename `EVMAddress` to `evmAddress` in `where` fields formatter
- refactor: remove `@ethersproject/address` as dependency, will use the formatter exported by snapshot.js

### How to tests

- `yarn link` this branch of the SDK https://github.com/snapshot-labs/snapshot.js/pull/1032
- `yarn dev`
- you should be able to query fields by starknet address (e.g using a non padded starknet address or checksummed address, address will always be converted to padded/lowercased address before sent to mysql)

### CODE REVIEW

File: src/graphql/helpers.ts [hunk 1]
```diff
@@ -31,7 +31,15 @@ export function formatAddress(address: string): string {
   try {
     return getAddress(address);
   } catch (error) {
-    // return same address for now, but return error in the future, if address is not EVM or sn address
+    // Pad starknet address to 64 characters
```
Comment (bonustrack) [inline src/graphql/helpers.ts:34]: Can we support Starknet address without going through `getAddress` function before?

Comment (wa0x6e) [reply src/graphql/helpers.ts:34]: Yes, I plan to move this function to the SDK, so we can reuse it inside sequencer and hub, with a more robust address detection

File: src/graphql/helpers.ts [hunk 2]
```diff
@@ -1,6 +1,13 @@
-import { getAddress } from '@ethersproject/address';
+import {
+  getFormattedAddress,
+  isEvmAddress,
+  isStarknetAddress
+} from '@snapshot-labs/snapshot.js/src/utils';
```
Comment (ChaituVR) [inline src/graphql/helpers.ts:5]: I think we should import like everywhere
Untitled 5

File: src/graphql/operations/user.ts [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
+import { formatUser, formatAddresses, PublicError } from '../helpers';
 
 export default async function (parent, args) {
-  const id = formatAddress(args.id);
+  const addresses = formatAddresses(
+    [args.id],
+    ['evmAddress', 'starknetAddress']
+  );
+  if (!addresses.length) throw new PublicError('Invalid address');
+
```
Comment (ChaituVR) [inline src/graphql/operations/user.ts:12]: Feel too confusing, it can't autoformat address.?

Comment (ChaituVR) [reply src/graphql/operations/user.ts:12]: i think we should let `formatAddresses` take care of types and errors

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/graphql/helpers.ts (hunk 2 above)
Comment (wa0x6e) [reply src/graphql/helpers.ts:5]: Weird, I don't have this error on the console. What version of node are you using ?

File: src/graphql/operations/user.ts (hunk 3 above)
Comment (wa0x6e) [reply src/graphql/operations/user.ts]: I we throw the error from `formatAddress`, we can not have a detailed error message explaining which field the error is coming from

File: src/graphql/helpers.ts [hunk 4]
```diff
@@ -124,27 +118,67 @@ export function formatSpace({
...
+        types.includes('starknetAddress') &&
+        snapshot.utils.isStarknetAddress(address)
+      ) {
+        results.push(snapshot.utils.getFormattedAddress(address, 'starknet'));
+      }
+
+      return results;
+    })
+    .flat()
+    .filter(Boolean);
```
Comment (ChaituVR) [inline src/graphql/helpers.ts:147]: I think it will be same with out these 🤔 
```suggestion
    });
```

File: src/graphql/helpers.ts [hunk 5]
```diff
@@ -124,27 +118,67 @@ export function formatSpace({
...
-      }
+
+      conditions.forEach(condition => {
+        const key = `${field}${condition}`;
+
+        if (!where[key]) return;
+
+        const formattedAddresses = uniq(
+          formatAddresses(castArray(where[key]), arrayType)
+        );
```
Comment (ChaituVR) [inline src/graphql/helpers.ts:171]: I think even if there is at least one invalid addresses, we should return error

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/graphql/helpers.ts (hunk 5 above)
Comment (ChaituVR) [reply src/graphql/helpers.ts:171]: maybe we can check length of where[key] and

**Review by ChaituVR — APPROVED** [review]
tAck

added few comments but not too important

**Pull request merged**